src/Orleans/Messaging/ProxiedMessageCenter.cs

```diff
@@ -425,6 +429,19 @@ internal void UpdateClientId(GrainId clientId)
             ClientId = clientId;
         }
 
+        internal void OnGatewayConnectionOpen()
+        {
+            Interlocked.Increment(ref numberOfConnectedGateways);
+        }
+
+        internal void OnGatewayConnectionClosed()
+        {
+            if (Interlocked.Decrement(ref numberOfConnectedGateways) == 0)
+            {
+                GrainClient.NotifyClusterConnectionLost();
```

Comment:
+1
May also want to consider exposing this signal in a thread safe way.. possibly via task completion?